pkg/kubelet/probe.go

```diff
@@ -24,53 +24,109 @@ import (
 	httprobe "github.com/GoogleCloudPlatform/kubernetes/pkg/probe/http"
 	tcprobe "github.com/GoogleCloudPlatform/kubernetes/pkg/probe/tcp"
-	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/exec"
-
 	"github.com/fsouza/go-dockerclient"
+
 	"github.com/golang/glog"
 )
 
 const maxProbeRetries = 3
 
-// probeContainer executes the given probe on a container and returns the result.
-// If the probe is nil this returns Success. If the probe's initial delay has not passed
-// since the creation of the container, this returns the defaultResult. It will then attempt
-// to execute the probe repeatedly up to maxProbeRetries times, and return on the first
-// successful result, else returning the last unsucessful result and error.
-func (kl *Kubelet) probeContainer(p *api.Probe,
-	podFullName string,
-	podUID types.UID,
-	status api.PodStatus,
-	container api.Container,
-	dockerContainer *docker.APIContainers,
-	defaultResult probe.Result) (probe.Result, error) {
+// probeContainer probes the liveness/readiness of the given container.
+// If the container's liveness probe is unsuccessful, set readiness to false.
+// If liveness is successful, do a readiness check and set readiness accordingly.
+func (kl *Kubelet) probeContainer(pod *api.BoundPod, status api.PodStatus, container api.Container, dockerContainer *docker.APIContainers) (probe.Result, error) {
+	// Probe liveness.
+	live, err := kl.probeContainerLiveness(pod, status, container, dockerContainer)
+	if err != nil {
+		glog.V(1).Infof("liveness probe errored: %v", err)
+		kl.readiness.set(dockerContainer.ID, false)
+		return probe.Unknown, err
+	}
+	if live != probe.Success {
+		glog.V(1).Infof("liveness probe unsuccessful: %v", live)
+		kl.readiness.set(dockerContainer.ID, false)
+		return live, nil
+	}
+
+	// Probe readiness.
+	ready, err := kl.probeContainerReadiness(pod, status, container, dockerContainer)
+	if err == nil && ready == probe.Success {
+		glog.V(1).Infof("readiness probe successful: %v", ready)
+		kl.readiness.set(dockerContainer.ID, true)
+		return probe.Success, nil
+	}
+
+	glog.V(1).Infof("readiness probe failed/errored: %v, %v", ready, err)
+	kl.readiness.set(dockerContainer.ID, false)
+
+	containerID := dockertools.DockerID(dockerContainer.ID)
+	ref, ok := kl.getRef(containerID)
+	if !ok {
+		glog.Warningf("No ref for pod '%v' - '%v'", containerID, container.Name)
+	} else {
+		kl.recorder.Eventf(ref, "unhealthy", "Liveness Probe Failed %v - %v", containerID, container.Name)
+	}
+	return ready, err
+}
+
+// probeContainerLiveness probes the liveness of a container.
+// If the initalDelay since container creation on liveness probe has not passed the probe will return probe.Success.
+func (kl *Kubelet) probeContainerLiveness(pod *api.BoundPod, status api.PodStatus, container api.Container, dockerContainer *docker.APIContainers) (probe.Result, error) {
```

Comment:
I think we can collapse these back into one function that takes the probe and the defaultResult. The code is near identical.